seastar/src/gate.rs

```diff
@@ -0,0 +1,136 @@
+use cxx::UniquePtr;
+use std::fmt;
+use std::future::Future;
+use std::pin::Pin;
+
+#[cxx::bridge(namespace = "seastar")]
+mod ffi {
+    unsafe extern "C++" {
+        include!("seastar/src/gate.hh");
+
+        type gate;
+        type gate_holder;
+
+        type CloseGateFuture = crate::CloseGateFuture;
+
+        fn new_gate() -> UniquePtr<gate>;
+        fn new_gate_holder(gate: &UniquePtr<gate>) -> Result<UniquePtr<gate_holder>>;
+        fn close_gate(gate: &UniquePtr<gate>) -> CloseGateFuture;
+    }
+}
+
+#[cxx_async::bridge(namespace = seastar)]
+unsafe impl Future for CloseGateFuture {
+    type Output = ();
+}
+
+use ffi::*;
+
+/// Error returned by [try_enter](Gate::try_enter) when called on closed gate.
+pub struct GateClosedError;
+
+impl fmt::Debug for GateClosedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "GateClosedError: gate closed")
+    }
+}
+
+/// Facility to stop new requests, and to tell when existing requests are done.
+///
+/// When stopping a service that serves asynchronous requests, we are faced with
+/// two problems: preventing new requests from coming in, and knowing when existing
+/// requests have completed. The [Gate] class provides a solution.
+pub struct Gate {
+    gate: UniquePtr<gate>,
+}
+
+/// Facility to hold a gate opened using RAII.
+///
+/// A [GateHolder] is obtained when [try_enter](Gate::try_enter) succeeds.
+///
+/// The [Gate] is left when the [GateHolder] is dropped.
+pub struct GateHolder<'a> {
+    _gate: &'a gate,
+    _holder: UniquePtr<gate_holder>,
+}
+
+impl<'a> Gate {
+    /// Creates a new gate.
+    ///
+    /// The underlying `gate` type has a non-trivial move constructor,
+    /// so we can't return it directly by value.
+    pub fn new() -> Pin<Box<Self>> {
+        Box::pin(Gate { gate: new_gate() })
+    }
```

Comment:
`struct Gate` holds UniquePtr to `gate` instead of holding it directly, so putting another `Box` creates a heap indirection - is it really necessary? It's possible that I don't understand something, but why not:
1. Return `Self` from this function - it looks like `UniquePtr` can be freely moved around, so no need to pin it, and it already pins the value inside, so we can't move it out: `https://docs.rs/cxx/latest/cxx/struct.UniquePtr.html#method.as_mut`. That would be my preferred approach.
2. Hold the gate directly in the struct, without UniquePtr - then we actually need to use `Pin<Box<Self>>`

Reply:
I'm not sure if approach 2 can work. The gate is of type `gate` which is an opaque type from C++ so it cannot be moved by value across the ffi. Therefore, we are forced to create it on the Rust side. I don't think it's possible (how are we going to call a C++ constructor?).

Approach 1 should be correct. I considered description of this issue as something that shouldn't be changed, but after putting `gate` in UniquePtr it makes sense to change return type to `Self`. I'll fix this in a moment.

Reply:
Imo issue description are absolutely subject to change - the aim of the project is to create an ergonomic API, so if you can improve the API then by all means do it.